File: TypeScript/src/session1.td/Nat.ts
import { FabriqueNaturels } from './FabriqueNaturels';
import { SemiAnneauUnitaireEuclidien } from '../hierarchie/SemiAnneauUnitaireEuclidien';

export interface Nat extends FabriqueNaturels<Nat>, SemiAnneauUnitaireEuclidien<Nat> {
	/*
	 * Sélecteur pour la définition par récurrence.
	 */
	estNul() : boolean;
	/*
	 * Décomposition d'un successeur.
	 */
	predecesseur() : Nat;
	/*
	 * Unités en 0, dizaines en un, etc.
	 */
 	chiffre(i : number) : number;
	/*
	 * Nombre de chiffres (aucun 0 superflu en tête). 
	 */
	taille() : number;

	/*
	 * Représentation par un int.
	 */
	val() : number;

	equals(x : Nat) : boolean;
}

File: Java/src/session1/demo1/v2/Nat.java
package session1.demo1.v2;

public interface Nat {
	default Nat somme(Nat x){
		return this.creer(this.getInt() + x.getInt());
	}
	int chiffre(int i);

	int taille();

	int getInt();
	
	Nat creer(int x);
	Nat creer(String chiffres);
}
